draft/celery-5-high-level-architecture.rst

```diff
@@ -0,0 +1,1900 @@
+:term:`message brokers <message broker>` or even clusters of
+:term:`message brokers <message broker>`.
+
+Data Sources and Sinks
+++++++++++++++++++++++
+
+Data Sources are a new concept in Celery.
+Data Sinks are a concept which replaces Result Backends.
+
+Data Sinks consume :term:`Document Messages <Document Message>` while Data Sources
+produce them.
+
+Data Sources
+~~~~~~~~~~~~
+
+Data Sources are :ref:`task <draft/celery-5-high-level-architecture:Services>`
+which either listen or poll for incoming data from a data source such as a
+database, a file system or an HTTP(S) endpoint.
+
+These services produce :term:`Document Messages <Document Message>`.
+
+Tasks which are subscribed to Data Sources will receive the raw document
+messages for further processing.
+
+.. admonition:: Example
+
+  We'd like to design a feature which locks Github issues immediately after
+  they are closed.
+
+  Github uses Webhooks to notify us when an issue is closed.
+
+  We set up a Data Source which starts an HTTPS server and expects incoming
+  HTTP requests on an endpoint.
+
+  Whenever a request arrives a :term:`Document Message` is published.
+
+Data Sinks
+~~~~~~~~~~
+
+A result from a :ref:`task <draft/celery-5-high-level-architecture:Tasks>`
+produces a :term:`Document Message` which a Data Sink or multiple Data Sinks
+consume.
+
+These :term:`Document Messages <Document Message>` are then stored in the Sinks
+the task is registered to.
+
+.. admonition:: Example
+
+  We have a task which calculates the hourly average impressions of a user's
+  post over a period of time.
+
+  The BI team requires the data to be inserted to `BigQuery <https://cloud.google.com/bigquery/>`_
+  because it uses it to research the effectiveness of users posts.
+
+  However, the user-facing post analytics dashboard also requires this data
+  and the team that maintains it doesn't want to use BigQuery because it is not
+  a cost-effective solution and because they already use `MongoDB <https://mongodb.com/>`_
+  to store all user-facing analytics data.
+
+  To resolve the issue we declare that the task routes it's results to two data
+  sinks. One for the BI team and the other for the analytics team.
+
+  Each data sink is configured to insert the data to a specific table
+  or collection.
+
+Controller
+----------
+
+The Controller is responsible for managing the lifecycle of all other Celery
+components.
+
+Celery 5 is a more complex system with multiple components and will often be
+deployed in high throughput, highly available production systems.
+
+The introduction of multiple components require us to have another component
+that manages the entire Celery cluster.
+
+During the lifecycle of a worker the Controller also manages and optimizes the
+execution of tasks to ensure we maximize the utilization of all our resources
+and to prevent expected errors.
+
+.. note::
+
+  The Controller is meant to be run as a user service.
```

Comment:
Although it is a logical constraint, and from a security perspective this should certainly be avoided, it may pose some problems, which I think have already been encountered in certain Github issue discussions:
- the more widespread use of Docker containers which use `root` as the user
- proper detection of running as a super-user is sometimes difficult and not always reliable

Perhaps this is mostly an implementation detail and also outside of Celery responsibility.

Reply:
Users should not run processes using the Docker's root user.
See [here](https://medium.com/@mccode/processes-in-containers-should-not-run-as-root-2feae3f0df3b) for why.
We do our best to detect and warn when using root privileges. If we cannot, I don't know what we should do.

Reply:
Just to clarify, I didn't mean that running processes as root in Docker containers is safe in any way. My point was that unfortunately it may happen more often. I agree that we should promote security best practices, but as I said before I am not sure we can reliably detect superuser privileges, e.g. what happens if the user is in sudoers? We could also check what Django or Flask do (or other server frameworks) with regard to security. Either way, I agree, we could just report with a warning :smile: .

Reply:
did any of you checked kata-containers?

Reply:
> did any of you check kata-containers?

Yes. A while ago. They are nice.